### PR TITLE
Add C++ Solution + Editorial to Snakes (Gold)

### PR DESCRIPTION
**Pull request by ryanchou-dev**
lmk if you want me to go into more detail with the editorial

---
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

File: solutions/usaco-945.mdx [hunk 1]
```diff
@@ -2,16 +2,71 @@
...
 **Time Complexity:** $\mathcal{O}(N^3)$
 
 <LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int MX_SZ = 400;
```
Comment (SansPapyrus683) [inline solutions/usaco-945.mdx:24]: MAX_N or MAX_SIZE
plz don't shorten the names

File: solutions/usaco-945.mdx [hunk 2]
```diff
@@ -2,16 +2,71 @@
...
+	vector<int> groups(n + 1);
+
+	for (int i = 1; i <= n; i++) {
+		cin >> groups[i];
+	}
+
+	// dp[i][j] = min sum of net sizes after catching first i snakes with j resizings
+	vector<vector<int>> dp(MX_SZ + 1, vector<int>(MX_SZ + 1, 0));
+
+	int lg = -1, spaceused = 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-945.mdx:42]: what the heck is `lg`
`spaceused` i kinda get but it's still up in the air

File: solutions/usaco-945.mdx [hunk 3]
```diff
@@ -2,16 +2,71 @@
...
+	cin >> n >> k;
+
+	vector<int> groups(n + 1);
+
+	for (int i = 1; i <= n; i++) {
+		cin >> groups[i];
+	}
+
+	// dp[i][j] = min sum of net sizes after catching first i snakes with j resizings
+	vector<vector<int>> dp(MX_SZ + 1, vector<int>(MX_SZ + 1, 0));
```
Comment (SansPapyrus683) [inline solutions/usaco-945.mdx:40]: vectors don't need constants
just use `n` & delete the constant

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-945.mdx (hunk 2 above)
Comment (ryanchou-dev) [reply solutions/usaco-945.mdx:42]: largest group (ill rename to make it more clear)

File: solutions/usaco-945.mdx [hunk 4]
```diff
@@ -2,16 +2,66 @@
...
+	
+	int n, k;
+	cin >> n >> k;
+
+	vector<int> groups(n + 1);
+	for (int i = 1; i <= n; i++) {
+		cin >> groups[i];
+	}
+
+	// dp[i][j] = min sum of net sizes after catching first i snakes with j resizings
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-945.mdx:36]:
```suggestion
	// dp[i][j] = the min sum of net sizes after catching first i snakes with j resizings
```

File: solutions/usaco-945.mdx [hunk 5]
```diff
@@ -2,16 +2,66 @@
...
+
+	int largest_group = -1;
+	int space_used = 0;
+	for (int i = 1; i <= n; i++) {
+		largest_group = max(largest_group, groups[i]);
+		
+		/*
+		 * base case: with zero resizings, Bessie must start off with
+		 * a net large enough to catch the largest group of snakes
+		*/
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-945.mdx:47]:
```suggestion
		// base case: with zero resizings, Bessie must start off with
		// a net large enough to catch the largest group of snakes
```

File: solutions/usaco-945.mdx [hunk 6]
```diff
@@ -2,16 +2,66 @@
...
+		dp[i][0] = largest_group * i;
+
+		for (int j = 1; j <= k; j++) {
+			dp[i][j] = INT32_MAX;
+			int cur = groups[i];
+			for (int c = i - 1; c >= 0; c--) {
+				// consider the case where we switch after collecting the first c groups
+				dp[i][j] = min(dp[i][j], dp[c][j - 1] + cur * (i - c));
+				cur = max(cur, groups[c]);
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-945.mdx:56]:
```suggestion
			int netSize = groups[i];
			for (int c = i - 1; c >= 0; c--) {
				// consider the case where we switch after collecting the first c groups
				dp[i][j] = min(dp[i][j], dp[c][j - 1] + netSize * (i - c));
				netSize = max(netSize , groups[c]);
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/usaco-945.mdx (hunk 5 above)
Comment (ryanchou-dev) [reply solutions/usaco-945.mdx:47]: imo multiline comments are neater

File: solutions/usaco-945.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/usaco-945.mdx]: ryan can you rename to net_size? think snake_case is convention for c++

Comment (ryanchou-dev) [reply solutions/usaco-945.mdx]: alright

**Review by SansPapyrus683 — APPROVED** [review]

**Review by TruaShamu — APPROVED** [review]

File: solutions/usaco-945.mdx [hunk 7]
```diff
@@ -2,16 +2,66 @@
...
+	
+	int n, k;
+	cin >> n >> k;
+
+	vector<int> groups(n + 1);
+	for (int i = 1; i <= n; i++) {
+		cin >> groups[i];
+	}
+
+	// dp[i][j] = the min sum of net sizes after catching first i snakes with j resizings
```
Comment (maggie-j-liu) [inline solutions/usaco-945.mdx:36]: i think this line and line 54 are too long

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Comment by ryanchou-dev** [comment]
**comments

**Review by maggie-j-liu — APPROVED** [review]

**Pull request merged**